### PR TITLE
Capability tracking

### PR DESCRIPTION
**Pull request by tmat**
Track capabilities required from the runtime in order to apply given update.

Follow-up pass to the debugger

### CODE REVIEW

**Comment by tmat** [comment]
@davidwengier PTAL. Probably best to review each commit separately.

**Review by davidwengier — APPROVED** [review]
Is it worth logging telemetry for which capabilities were missing, now that we're tracking it? Might be easier than extrapolating it from the existing rude edit telemetry.

File: src/Features/Core/Portable/EditAndContinue/EditAndContinueCapabilitiesGrantor.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.CodeAnalysis.EditAndContinue
+{
+    /// <summary>
+    /// Grants capabilities.
```
[the remark this answers is not in the document]
Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/EditAndContinueCapabilitiesGrantor.cs:8]: I also struggled with the name but this is best I was able to come up with.

**Comment by tmat** [comment]
> Is it worth logging telemetry for which capabilities were missing, now that we're tracking it? Might be easier than extrapolating it from the existing rude edit telemetry.

We are tracking those that were granted but not those that were denied. We immediately report a rude edit when a required capability is not granted. The rude edit should clearly indicate (to the user as well as in the existing telemetry) what operation can't be performed due to runtime not having the capability. So it seems we have all info we need in the telemetry. Do you have a specific telemetry based question on mind that can't be (or is hard to get) answered today?

**Comment by davidwengier** [comment]
> Do you have a specific telemetry based question on mind that can't be (or is hard to get) answered today?

Nothing specific I'm thinking of, just wondering if it's worth being able to say to the runtime team "X hot reload attempts failed because AddNewTypeDef wasn't supported", as a general statement, because rude edits are obviously going to be more granular.

**Comment by tmat** [comment]
I see. I think as long as we include "NotSupportedByRuntime" in rude edit ids it should be pretty easy to extract the data.

**Pull request merged**